### PR TITLE
Improve entity stake calculation in v2

### LINKED ISSUE #7663: Improve Entity Stake Calculation In V2
### Problem

The entity stake calculation in v2 is very slow and brings the entire system to a crawl.

### Solution

Optimize query to be performant in v2

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by xin-hedera**
**Description**:

This PR improves entity stake calculation in v2

- Change `entity_stake_temp` to be a permanent table in temp schema, distribute the the table and colocate it with `entity_stake` in v2
- Change`entity_state_start` to be a permanent table in temp schema, distribute the table and colocate it with `entity` in v2
- Break down the SQL for creating `entity_state_start` to minimize the number of timestamp partitions scanned
- Optimize the SQL query for `EntityStakeRepository.updated()`

**Related issue(s)**:

Fixes #7663

**Notes for reviewer**:

Tested with citus db with the range of crypto transfers to scan limited to 600s

- `createEntityStateStart` takes about 30s when cold (all data read from disk), 2s when hot (the second run, most data it needs is in memory0
- `updateEntityStake` takes about 20s for ~54000 entities

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepository.java [hunk 1]
```diff
@@ -150,8 +43,12 @@ select coalesce((select end_stake_period from entity_stake where id = 800), -1)
...
               select (select id from entity where id = 800) as account_id
             )
-            select account_id is null or epoch_day = -1 or epoch_day = end_stake_period
-            from last_epoch_day, entity_stake_info, staking_reward_account
+            select case when account_id is null then true
+                        when (select epoch_day from last_epoch_day)
+                          in (-1, (select end_stake_period from entity_stake_info)) then true
+                        else false
+                   end
+            from staking_reward_account
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepository.java:51]: this is faster in v2, 10ms v.s. 50ms

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepository.java [hunk 2]
```diff
@@ -232,14 +135,13 @@ insert into entity_stake_temp (end_stake_period, id, pending_reward,
...
-                 when coalesce(staked_node_id_start, -1) = -1
-                      then 0
+                 when coalesce(staked_node_id_start, -1) = -1 then 0
                  when reward_rate is null then pending_reward
                  when ess.stake_period_start > epoch_day - 1 then 0
                  when ess.stake_period_start = epoch_day - 1 then reward_rate * stake_total_start_whole_bar
                  when epoch_day - ess.stake_period_start > 365
                    then pending_reward + reward_rate * stake_total_start_whole_bar
-                     - coalesce((select reward_rate from node_stake where epoch_day = ep.epoch_day - 365 and node_id = staked_node_id_start), 0) * stake_total_start_whole_bar
+                     - coalesce((select reward_rate from forfeited_period where node_id = staked_node_id_start), 0) * stake_total_start_whole_bar
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepository.java:144]: this part causes error with citus: ERROR:  direct joins between distributed and local tables are not supported

so have to pull it to the `forfeited_period` CTE.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/EntityStakeRepositoryTest.java [hunk 3]
```diff
@@ -275,10 +275,7 @@ void createEntityStateStartWhenEmptyAccountBalance() {
                 .nodeStake()
                 .customize(ns -> ns.consensusTimestamp(timestamp).epochDay(epochDay))
                 .persist();
-        domainBuilder
-                .entity()
-                .customize(e -> e.timestampRange(Range.atLeast(timestamp - 5000L)))
-                .persist();
+        domainBuilder.entity(STAKING_REWARD_ACCOUNT, timestamp - 5000L).persist();
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/repository/EntityStakeRepositoryTest.java:278]: fix the test case set up to increase test coverage

File: hedera-mirror-importer/src/main/resources/db/migration/common/R__01_temp_tables.sql [hunk 4]
```diff
@@ -17,6 +19,7 @@ create unlogged table if not exists ${tempSchema}.contract_state_temp as table c
 create unlogged table if not exists ${tempSchema}.crypto_allowance_temp as table crypto_allowance limit 0;
 create unlogged table if not exists ${tempSchema}.custom_fee_temp as table custom_fee limit 0;
 create unlogged table if not exists ${tempSchema}.dissociate_token_transfer as table token_transfer limit 0;
+create unlogged table if not exists ${tempSchema}.entity_stake_temp as table entity_stake limit 0;
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/resources/db/migration/common/R__01_temp_tables.sql:22]: note the indexes for the two tables added here are not created like other tables. Reason is it's more performant to first insert data, then create the index, especially when there's a lot of data to insert.

I created a ticket to track it.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepositoryCustomImpl.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.transaction.annotation.Transactional;
+
+@Named
+@RequiredArgsConstructor
+class EntityStakeRepositoryCustomImpl implements EntityStakeRepositoryCustom {
+
+    private static final String CLEANUP_TABLE_SQL =
+            """
+            drop index if exists entity_state_start__id;
```
Comment (jnels124) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/repository/EntityStakeRepositoryCustomImpl.java:36]: any benefit to dropping the index each time and recreating vs creating once in the repeatable migration? Guess this doesn't run often enough to matter much here but curious as we have other temp tables where we could benefit from the approach.

**Review by edwin-greene — APPROVED** [review]
Looks good

**Review by jnels124 — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**